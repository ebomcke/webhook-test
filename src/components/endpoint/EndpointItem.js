import React, { Component } from 'react';
import { Feed, Segment, Icon, Modal } from 'semantic-ui-react';
import { compose } from 'recompose';
import { withNavigation } from '../../contexts/Navigation';
import Timestamp from '../base/Timestamp';
import { withFirestore } from 'react-firestore';
import { withAccount } from '../../contexts/Account';
import WebhookTester from '../webhook/WebhookTester';

class EndpointItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      testWebhookModalOpen: false,
    };
  }

  handleOpen = () => this.setState({ testWebhookModalOpen: true });

  handleClose = () => this.setState({ testWebhookModalOpen: false });

  openEndpoint = () => {
    this.props.navigate({
      name: 'Endpoint',
      params: {
        endpointId: this.props.endpoint.id,
      },
    });
  }

  shortResponse = () => {
    const { endpoint } = this.props;
    let response = JSON.stringify(
      JSON.parse(endpoint.defaultResponse.body),
      null,
      2,
    );
    let lines = [`HTTP ${endpoint.defaultResponse.statusCode}:`].concat(
      response.split(/\n/),
    );
    if (lines.length > 6) {
      lines = lines.slice(0, 6).concat('[...]', lines.slice(-1));
    }
    return lines.join('\r\n');
  }

  deleteEndpoint = async () => {
    const { endpoint, firestore } = this.props;
    await firestore
      .collection('endpoints')
      .doc(endpoint.id)
      .delete();
  }

  render() {
    const { endpoint, account } = this.props;
    return (
      <Feed.Event className="endpoint-feed-item">
        <Feed.Label>
          <Icon name="at" />
        </Feed.Label>
        <Feed.Content>
          <Feed.Summary>
            <a name='open-endpoint-link' onClick={this.openEndpoint}>{endpoint.path}</a>
            <Feed.Date>
              <Timestamp timestamp={endpoint.lastActive} prefix="Last active" />
            </Feed.Date>
          </Feed.Summary>
          <Feed.Extra>
            <p>
              URL:{' '}
              {`${process.env.API_URL}/webhook/${account.organisationName}/${
                endpoint.path
              }`}
            </p>
          </Feed.Extra>
          <Feed.Extra text>
            <div>Response:</div>
            <Segment inverted>
              <pre>
                <code>{this.shortResponse()}</code>
              </pre>
            </Segment>
          </Feed.Extra>
          <Feed.Meta>
            <a onClick={this.deleteEndpoint}>
              <Icon name="remove" />
              Remove
            </a>
            <Modal
              open={this.state.testWebhookModalOpen}
              onClose={this.handleClose}
              trigger={
                <a name="send-test-message-link" onClick={this.handleOpen}>
                  <Icon name="send" />
                  Send test message
                </a>
              }
            >
              <Modal.Header>Send test message</Modal.Header>
              <Modal.Content>
                <WebhookTester
                  endpoint={endpoint}
                  onComplete={this.handleClose}
                />
              </Modal.Content>
            </Modal>
          </Feed.Meta>
        </Feed.Content>
      </Feed.Event>
    );
  }
}

export default compose(
  withNavigation,
  withFirestore,
  withAccount,
)(EndpointItem);
