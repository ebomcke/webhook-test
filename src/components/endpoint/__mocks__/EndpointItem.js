import React from 'react';

export default props => (
  <div>
    This is a placeholder for the endpoint item
    <pre>
      <code>{JSON.stringify(props)}</code>
    </pre>
  </div>
);
