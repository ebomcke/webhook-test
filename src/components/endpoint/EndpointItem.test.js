import React from 'react';
import { create } from 'react-test-renderer';
import EndpointItem from './EndpointItem';

jest.mock('../base/Timestamp');

const endpoint = {
  id: 'endpoint',
  path: 'unit-test',
  lastActive: 'last-active',
  defaultResponse: {
    statusCode: 200,
    body: JSON.stringify({ unitTest: true }),
  },
};

describe('Rendering tests', () => {
  test('Response under threshold should be displayed in full', () => {
    const endpointItem = create(<EndpointItem endpoint={endpoint} />);
    expect(endpointItem.toJSON()).toMatchSnapshot();
  });

  test('Response over threshold should be shortened', () => {
    const endpointItem = create(
      <EndpointItem
        endpoint={Object.assign({}, endpoint, {
          defaultResponse: {
            body: JSON.stringify({
              first: 'line',
              second: 'line',
              third: 'line',
              fourth: 'line',
              fifth: 'line',
              sixth: 'line',
              seventh: 'line',
            }),
          },
        })}
      />,
    );
    expect(endpointItem.toJSON()).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  test('Clicking on the endpoint path should navigate to the endpoint view', () => {
    const navigate = jest.fn();
    const endpointItem = create(
      <EndpointItem endpoint={endpoint} navigate={navigate} />,
    );
    const openDetailsLink = endpointItem.root.findByProps({
      name: 'open-endpoint-link',
    });
    openDetailsLink.props.onClick();
    expect(navigate.mock.calls.length).toBe(1);
    expect(navigate.mock.calls[0][0]).toEqual({
      name: 'Endpoint',
      params: {
        endpointId: endpoint.id,
      },
    });
  });
});
