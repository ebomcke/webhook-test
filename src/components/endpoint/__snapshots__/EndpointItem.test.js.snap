// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering tests Response over threshold should be shortened 1`] = `
<div
  className="event endpoint-feed-item"
>
  <div
    className="label"
  >
    <i
      aria-hidden="true"
      className="at icon"
    />
  </div>
  <div
    className="content"
  >
    <div
      className="summary"
    >
      <a
        name="open-endpoint-link"
        onClick={[Function]}
      >
        unit-test
      </a>
      <div
        className="date"
      >
        <div>
          This is a placeholder for the Timestamp.
          <pre>
            <code>
              {"timestamp":"last-active","prefix":"Last active"}
            </code>
          </pre>
        </div>
      </div>
    </div>
    <div
      className="extra"
    >
      <p>
        URL:
         
        https://webhook-tester-1547141127436.appspot.com/webhook/unit-test-org/unit-test
      </p>
    </div>
    <div
      className="text extra"
    >
      <div>
        Response:
      </div>
      <div
        className="ui inverted segment"
      >
        <pre>
          <code>
            HTTP undefined:
{
  "first": "line",
  "second": "line",
  "third": "line",
  "fourth": "line",
[...]
}
          </code>
        </pre>
      </div>
    </div>
    <div
      className="meta"
    >
      <a
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="remove icon"
        />
        Remove
      </a>
      <a
        name="send-test-message-link"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="send icon"
        />
        Send test message
      </a>
    </div>
  </div>
</div>
`;

exports[`Rendering tests Response under threshold should be displayed in full 1`] = `
<div
  className="event endpoint-feed-item"
>
  <div
    className="label"
  >
    <i
      aria-hidden="true"
      className="at icon"
    />
  </div>
  <div
    className="content"
  >
    <div
      className="summary"
    >
      <a
        name="open-endpoint-link"
        onClick={[Function]}
      >
        unit-test
      </a>
      <div
        className="date"
      >
        <div>
          This is a placeholder for the Timestamp.
          <pre>
            <code>
              {"timestamp":"last-active","prefix":"Last active"}
            </code>
          </pre>
        </div>
      </div>
    </div>
    <div
      className="extra"
    >
      <p>
        URL:
         
        https://webhook-tester-1547141127436.appspot.com/webhook/unit-test-org/unit-test
      </p>
    </div>
    <div
      className="text extra"
    >
      <div>
        Response:
      </div>
      <div
        className="ui inverted segment"
      >
        <pre>
          <code>
            HTTP 200:
{
  "unitTest": true
}
          </code>
        </pre>
      </div>
    </div>
    <div
      className="meta"
    >
      <a
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="remove icon"
        />
        Remove
      </a>
      <a
        name="send-test-message-link"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="send icon"
        />
        Send test message
      </a>
    </div>
  </div>
</div>
`;
